File: api/routers/cx.py
"""GET /api/cx — Pilier Expérience Client."""
from __future__ import annotations

from collections import Counter

from fastapi import APIRouter

from ..loaders import load_ai_latest, load_excel_cx, load_store_latest
from ..kpis import (
    enchantements_from_records,
    irritants_from_records,
    nps_proxy,
    rating_by_month,
    rating_distribution,
)

router = APIRouter()


@router.get("/cx")
def get_cx():
    ai = load_ai_latest()
    review_enriched = ai.get("review", [])
    store_reviews = load_store_latest()
    excel_cx = load_excel_cx()

    # Normalize excel_cx: field names differ from ai_runs schema
    for r in excel_cx:
        if not r.get("published_at") and r.get("date"):
            r["published_at"] = r["date"]
        if not r.get("sentiment_label") and r.get("sentiment"):
            r["sentiment_label"] = r["sentiment"].lower()

    # Exclude employee partition — only customer + store reviews for CX
    cx_enriched = [
        r for r in review_enriched
        if r.get("source_partition") not in ("employee",)
    ]

    # Combine all review-type records
    all_reviews = store_reviews + cx_enriched + excel_cx

    # ── KPIs ────────────────────────────────────────────────────────────────
    ratings = []
    for r in all_reviews:
        try:
            v = float(r.get("rating") or r.get("note") or 0)
            if 1 <= v <= 5:
                ratings.append(v)
        except (ValueError, TypeError):
            pass

    avg_rating = round(sum(ratings) / len(ratings), 2) if ratings else 0.0
    nps = nps_proxy(all_reviews)

    # SAV negative pct — records mentioning SAV themes that are negative
    sav_themes = {"service_client", "retour_remboursement"}
    sav_neg = sum(
        1 for r in all_reviews
        if r.get("sentiment_label") == "negative"
        and any(t in sav_themes for t in (r.get("themes") or []))
    )
    sav_neg_pct = round(sav_neg / len(all_reviews), 3) if all_reviews else 0.0

    # ── Rating par mois ──────────────────────────────────────────────────────
    rbm = rating_by_month(all_reviews)

    # ── Distribution des notes ───────────────────────────────────────────────
    dist = rating_distribution(all_reviews)

    # ── Irritants & Enchantements ─────────────────────────────────────────────
    # Use enriched records with themes — exclude employee partition
    enriched = cx_enriched + [r for r in store_reviews if r.get("themes")]
    irritants = irritants_from_records(enriched, top_n=5)
    enchantements = enchantements_from_records(enriched, top_n=3)

    # ── Sources (with URLs) ───────────────────────────────────────────────────
    source_data: dict[str, dict] = {}
    for r in all_reviews:
        name = (
            r.get("source_name")
            or r.get("platform")
            or r.get("site")
            or r.get("_sheet")
            or "Autre"
        )
        if name not in source_data:
            source_data[name] = {
                "count": 0,
                "url": r.get("source_url") or r.get("google_maps_url") or None,
            }
        source_data[name]["count"] += 1
        # Keep first non-null URL found
        if not source_data[name]["url"]:
            source_data[name]["url"] = r.get("source_url") or r.get("google_maps_url") or None

    sources = [
        {"name": k, "count": v["count"], "url": v["url"]}
        for k, v in sorted(source_data.items(), key=lambda x: -x[1]["count"])[:8]
    ]

    return {
        "kpis": {
            "avg_rating": avg_rating,
            "nps_proxy": nps,
            "total_reviews": len(all_reviews),
            "sav_negative_pct": sav_neg_pct,
        },
        "rating_by_month": rbm,
        "rating_distribution": dist,
        "irritants": irritants,
        "enchantements": enchantements,
        "sources": sources,
    }

File: server/routes/cx.mjs
import { Router } from 'express';
import { getDb, parseJsonCol } from '../db.mjs';
import { enchantements, irritants, npsProxy, ratingByMonth, ratingDistribution } from '../kpis.mjs';

const router = Router();

router.get('/cx', (req, res) => {
  const db = getDb();

  const reviewEnriched = db.prepare("SELECT * FROM review_enriched WHERE source_partition != 'employee'")
    .all().map(r => parseJsonCol(r, 'themes'));
  const storeReviews = db.prepare('SELECT * FROM store_reviews').all().map(r => parseJsonCol(r, 'themes'));
  const excelCx = db.prepare('SELECT * FROM excel_cx').all().map(r => ({
    ...r,
    published_at: r.date || '',
    sentiment_label: (r.sentiment || '').toLowerCase(),
    themes: [],
  }));

  const allReviews = [...storeReviews, ...reviewEnriched, ...excelCx];

  // Avg rating
  const ratings = [];
  for (const r of allReviews) {
    const v = parseFloat(r.rating || r.note || 0);
    if (v >= 1 && v <= 5) ratings.push(v);
  }
  const avgRating = ratings.length ? Math.round(ratings.reduce((a, b) => a + b, 0) / ratings.length * 100) / 100 : 0;
  const nps = npsProxy(allReviews);

  // SAV negative pct
  const savThemes = new Set(['service_client', 'retour_remboursement']);
  const savNeg = allReviews.filter(r =>
    r.sentiment_label === 'negative' && (r.themes || []).some(t => savThemes.has(t))
  ).length;
  const savNegPct = allReviews.length ? Math.round(savNeg / allReviews.length * 1000) / 1000 : 0;

  const rbm = ratingByMonth(allReviews);
  const dist = ratingDistribution(allReviews);

  const enriched = [...reviewEnriched, ...storeReviews.filter(r => r.themes?.length)];
  const irr = irritants(enriched, 5);
  const ench = enchantements(enriched, 3);

  // Sources
  const sourceData = {};
  for (const r of allReviews) {
    const name = r.source_name || r.platform || r.site || 'Autre';
    if (!sourceData[name]) sourceData[name] = { count: 0, url: null };
    sourceData[name].count++;
    if (!sourceData[name].url) sourceData[name].url = r.source_url || r.google_maps_url || null;
  }
  const sources = Object.entries(sourceData)
    .sort((a, b) => b[1].count - a[1].count)
    .slice(0, 8)
    .map(([name, v]) => ({ name, count: v.count, url: v.url }));

  res.json({
    kpis: { avg_rating: avgRating, nps_proxy: nps, total_reviews: allReviews.length, sav_negative_pct: savNegPct },
    rating_by_month: rbm,
    rating_distribution: dist,
    irritants: irr,
    enchantements: ench,
    sources,
  });
});

export default router;
